ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,235 @@
+import os, shutil, time, json
+import contextlib2
+import tensorflow as tf
+from pathlib import Path
+from random import shuffle
+from datetime import datetime
+from ravenml.data.interfaces import CreateInput
+from ravenml.utils.question import cli_spinner, cli_spinner_wrapper
+from ravenml.utils.config import get_config
+from ravenml.data.helpers import default_filter, default_load_image_ids, copy_data_locally
+
+class DatasetWriter(object):
+
+    def __init__(self, create: CreateInput, **kwargs):
+        self.associated_files = kwargs['associated_files']
+        self.related_data_prefixes = kwargs['related_data_prefixes']
+        
+        metadata = create.metadata
+        self.num_folds = create.kfolds # Never actually used
+        self.test_percent = create.test_percent
+        self.label_to_int_dict = {}
+        self.image_ids = None
+        self.filter_metadata = None
+        self.temp_dir = create.plugin_metadata['temp_dir_path']
+        self.dataset_path = create.dataset_path
+        self.dataset_name = metadata['dataset_name']
+        self.created_by = metadata['created_by']
+        self.comments = metadata['comments']
+        self.plugin_name = create.plugin_metadata['architecture']
+        self.imageset_paths = create.imageset_paths
+        self.filter = metadata['filter']
+        self.cache = create.plugin_cache
+    
+    def construct_all(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def export_data(self, *args, **kwargs):
+        raise NotImplementedError
+    
+    def write_additional_files(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def filter_sets(self, metadata_prefix):
```

Comment:
nit: a better name might be `interactive_filter`.

Also, it might be a bit cleaner to have the decision on whether or not to perform filtering happen outside of the DatasetWriter, in the plugin's `core.py`. It's a couple lines of extra boilerplate, but it makes it so this method has a clearer API -- if you call `dataset_writer.interactive_filter()`, then it's guaranteed to interactively filter. Otherwise, the method should be renamed to `maybe_interactive_filter()`.